### PR TITLE
Modernize CMake

### PR DESCRIPTION
**Pull request by SimeonEhrig**
- add a vikunja CMake target, which can be linked
  - allows also the usage of `add_subdirectory`
- add CMake instruction to make it installable and findable via `find_package`

### CODE REVIEW

**Comment by SimeonEhrig** [comment]
@j-stephan Can you please review it, because you added the CMake install instruction in alpaka

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt:9]: MSVC screams.

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt:15]: @sbastrakov and @psychocoderHPC: is this how we expected users of alpaka to deal with our disabling of all backends by default? Because I was surprised back then when no backend was enabled by default in alpaka and someone of you told me, it's better that the user explicitely enables one. So vikunja is now enabling a backend automatic again. Is that good or bad now? I really do not know :S

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt:37]: So this is checking if vikunja is used as a subdirectory?

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt:78]: There was a talk on CMake at CppCon by Robert Schumacher (https://www.youtube.com/watch?v=sBP17HQAQjk) where he complained that some libraries take longer to build whether or not doxygen is installed or now. He recommended to at least give an option to disable building the documentation. Personally, I would not even build the documentation by default.

File: test/CMakeLists.txt
Comment (bernhardmgruber) [inline test/CMakeLists.txt:23]: Why are these `SYSTEM` includes?

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt:8]: That suggests C++14 for all subprojects. To enforce it, you also have to set `CMAKE_CXX_STANDARD_REQUIRED`.

Recently, I have also become a fan of `target_compile_features(${PROJECT_NAME} INTERFACE cxx_std_17)` for interface libraries, because you just state that the library needs C++17 and any project downstream can request something higher.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:9]: No problem, I also scream, if I use the MSVC XD Joke aside, I didn't look on this line. In general, the library is not tested with Windows or MacOSX. But before fixing this line, I should look after the reason. But maybe in another PR. Jiri's students need this stuff.

File: CMakeLists.txt
Comment (sbastrakov) [reply CMakeLists.txt:15]: This is a good question. This line enables alpaka sequential accelerator that should be available everywhere and so it's safe to do it. However this accelerator should never be used for "actual" (parallel) alpaka-based computing.

Comment (SimeonEhrig) [reply CMakeLists.txt:15]: I think, we should avoid it in general. But in this special case, I recognize, that the place is simply wrong. The backend is necessary for the tests and example, so I should move it to the CMakeLists.txt of it.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:37]: yes

File: test/CMakeLists.txt
Comment (SimeonEhrig) [reply test/CMakeLists.txt:23]: I believe, it has something to do, if you use `#include <header.h>` instead `#include "header.h"` in your project, but I can check, if it is necessary. Maybe it was necessary, because the vikunja headers was direct include via cmake variable, instead using a target.

File: CMakeLists.txt
Comment (j-stephan) [reply CMakeLists.txt:9]: Is there a reason why we are setting`CMAKE_CXX_FLAGS` directly? That is usually considered bad style since this is a flag that is set by the user (if required). The only exception I know of is the (outdated) `FindCUDA.cmake` script which doesn't play nice with CMake targets. If vikunja isn't using this I suggest the following change:

```suggestion
if(MSVC)
    target_compile_options(vikunja INTERFACE "/W4")
else()
    target_compile_options(vikunja INTERFACE "-Wall -Wextra -pedantic")
endif()
```

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:17]: You can also specify a version here if required. This way the build system can detect incompatible versions if vikunja is used together with cupla.

```cmake
find_package(alpaka [version] [EXACT] REQUIRED)
```

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:50]: Why are we installing alpaka here? What if I already have an installed alpaka on my system?

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:84]: Shouldn't this be configured by the user?

File: test/CMakeLists.txt
Comment (j-stephan) [reply test/CMakeLists.txt:23]: This looks shaky. What if I point to an installed alpaka on my system? There shouldn't be a `test` include folder there.

**Review by j-stephan — COMMENTED** [review]
A few comments:

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:9]: It is an artifact, from the old CMake 2.x style without targets, which I forgot. Maybe the heading of the PR is wrong. My target was to make vikunja nice usable in other project. So didn't care about the sections, which are not direct required for my task. But I think, I should change my opinion ;-)

You suggestion sounds good. But I want to check also, if it necessary in general.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:50]: Because alpaka is a header-only library, it does not install any file. This needs to by another command (at least in the solution, which I found). The command is necessary to export the target definition. Otherwise the following code does not work:

```
target_link_libraries(${PROJECT_NAME}
  INTERFACE
  alpaka::alpaka)
  ```  
If you have a better solution, please tell me it.

File: test/CMakeLists.txt
Comment (j-stephan) [reply test/CMakeLists.txt:23]: Yes, I know. But you are pointing to alpaka's `test` directory, right? What if there is no `test` directory because I pointed to an alpaka installation?

File: CMakeLists.txt
Comment (bernhardmgruber) [reply CMakeLists.txt:84]: I think this just enables the testing capabilities of CMake and not the tests per se. CMake should then have an option `BUILD_TESTING` which allows to enable/disable building tests.

File: CMakeLists.txt
Comment (j-stephan) [reply CMakeLists.txt:50]: `alpaka::alpaka` should be already available if `find_package(alpaka)` was successful. Or am I missing something?

Comment (SimeonEhrig) [reply CMakeLists.txt:50]: Could be :thinking: I added `find_library("alpaka")` in the config file afterwards. This could solve the problem. I test it.

Comment (SimeonEhrig) [reply CMakeLists.txt:50]: The command does not install alpaka. It's necessary for the target configuration. I tested it with a alpaka, which is installed in my environment.

If I remove the `alpaka`, I get the following error: `CMake Error: install(EXPORT "vikunjaTargets" ...) includes target "vikunja" which requires target "alpaka" that is not in any export set.`

Comment (j-stephan) [reply CMakeLists.txt:50]: Interesting, I'll look into this. I think this may be a bug in alpaka's CMake scripts.

File: test/CMakeLists.txt
Comment (sliwowitz) [reply test/CMakeLists.txt:23]: Lines 21-23 seem like some leftovers, the test compile without them.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt]: I had a mistake in my mind. I thought it reduce the amount of warnings. I used @j-stephan approach and added a option to make it optional.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:20]: I forgot, that the acc cannot be chosen after `find_package`. Now I enable the acc in the top level CMakeLists.txt, if the build the of examples and/or tests are enabled and `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE` was not set before.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt]: I made it optional

File: test/CMakeLists.txt
Comment (SimeonEhrig) [reply test/CMakeLists.txt]: Removed `SYSTEM` and the lines 21-23.

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt]: I added the option `BUILD_TESTING`

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:66]: Do you want to test it, before merging or should I keep it. It looks ugly, but I didn't saw any malicious side effect.

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:14]: Suggested change:

```suggestion
option(BUILD_TESTING "Build the testing tree." OFF)
include(CTest)
```

CTest listens to the `BUILD_TESTING` option (defaults to `ON`, so we still need to override) and gives us the nice `ctest` command line interface. Since we already use `add_test` in the `test` directory this comes without additional work.

File: CMakeLists.txt
Comment (j-stephan) [reply CMakeLists.txt:66]: You can keep it for now, I don't know when I'll find the time. If I can solve this in alpaka the fix is very easily ported to vikunja.

**Comment by SimeonEhrig** [comment]
@j-stephan @bernhardmgruber 
Can you please approve, if the PR is ready for merge

**Comment by SimeonEhrig** [comment]
> @j-stephan @bernhardmgruber
> Can you please approve, if the PR is ready for merge

Please wait with approving. I found a horrible bug. I believe nesting `alpaka` is not possible. If I write

```cmake
set(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE ON)
find_package(vikunja)
```

or set `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE` via argument, it is not pass throw to the `find_package(alpaka)` in the `vikunja-config.cmake` and the acc is not enabled. Any ideas? Otherwise I have to remove alpaka from vikunja, in the user have to handle alpaka manual.

**Review by bernhardmgruber — APPROVED** [review]

**Comment by bernhardmgruber** [comment]
> > @j-stephan @bernhardmgruber
> > Can you please approve, if the PR is ready for merge
> 
> Please wait with approving. I found a horrible bug. I believe nesting `alpaka` is not possible. ...

I let you solve that with @sbastrakov or @psychocoderHPC. I think I would just not set `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE` myself but expect the user to set this on the cmake command line. But I do not mind how you solve that.

The rest is fine I think!

**Comment by SimeonEhrig** [comment]
> > > @j-stephan @bernhardmgruber
> > > Can you please approve, if the PR is ready for merge
> > 
> > 
> > Please wait with approving. I found a horrible bug. I believe nesting `alpaka` is not possible. ...
> 
> I let you solve that with @sbastrakov or @psychocoderHPC. I think I would just not set `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE` myself but expect the user to set this on the cmake command line. But I do not mind how you solve that.
> 
> The rest is fine I think!

In this case, the code was from my own example. Nothing for the public. In vikunja itself, the `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE` is only enabled by default, if you build the examples or the tests. And if you install vikunja as library, it does not set any acc.

**Comment by j-stephan** [comment]
Just to make sure I understand this correctly: `cmake -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE=ON /path/to/vikunja` does not enable the serial accelerator?

**Comment by SimeonEhrig** [comment]
> Just to make sure I understand this correctly: `cmake -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE=ON /path/to/vikunja` does not enable the serial accelerator?

Yes, if you mean `/path/to/vinkunja` is project, which uses vikunja via `find_package(vikunja)`. But I found really interesting detail. At the cmake configure phase, there is the message, that the acc is enabled:

```bash
$ cmake -GNinja .. -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE=ON
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Boost: /opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.74.0-lqyr7ou7iafiwvq5fb6ipp62o6l2gqlf/lib/cmake/Boost-1.74.0/BoostConfig.cmake (found suitable version "1.74.0", minimum required is "1.65.1")  missing components: fiber
-- ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sehrig/programming/vikunja/rand/build
$ cmake --build .
[1/2] Building CXX object CMakeFiles/vikRand.dir/main.cpp.o
FAILED: CMakeFiles/vikRand.dir/main.cpp.o 
/usr/bin/c++ -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=30 -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DBOOST_ALL_NO_LIB -isystem /home/sehrig/application/miniconda3/envs/vikunja/include -Wall -Wextra -pedantic -MD -MT CMakeFiles/vikRand.dir/main.cpp.o -MF CMakeFiles/vikRand.dir/main.cpp.o.d -o CMakeFiles/vikRand.dir/main.cpp.o -c ../main.cpp
../main.cpp: In function ‘int main()’:
../main.cpp:11:29: error: ‘AccCpuSerial’ in namespace ‘alpaka::acc’ does not name a template type
   11 |   using TAcc = alpaka::acc::AccCpuSerial<alpaka::dim::DimInt<3u>, std::uint64_t>;
      |                             ^~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```
I see, that the `-DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED` is missing in the compiler command. 

Here are the `CMakeLists.txt` and `main.cpp`

```cmake
cmake_minimum_required(VERSION 3.15)

set(_TARGET_NAME vikRand)
project(${_TARGET_NAME})

#find_package(alpaka REQUIRED)

find_package(vikunja REQUIRED)

alpaka_add_executable(${_TARGET_NAME})
target_sources(${_TARGET_NAME}
  PRIVATE
  main.cpp)
target_link_libraries(${_TARGET_NAME}
  PRIVATE
  #alpaka::alpaka
  vikunja::vikunja
  )
```

```c++
#include <alpaka/alpaka.hpp>
#include <vikunja/transform/transform.hpp>

#include <iostream>

int main(){

  using TAcc = alpaka::acc::AccCpuSerial<alpaka::dim::DimInt<3u>, std::uint64_t>;


  // data type of the data, which should be transformed
  using TData = float;
  //using Tix = alpaka::idx::Idx<TAcc>;
  //using Vec = alpaka::vec::Vec<TAcc>;

  std::cout << "Hello Vikunja" << std::endl;

  return 0;
}
```

**Comment by psychocoderHPC** [comment]
> ```cmake
> set(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE ON)
> ```

You can not set options like it is a normal variable. `set(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE ON)` will not work.
You must do it the ugly way: https://github.com/alpaka-group/cupla/blob/22dfdf4574d529b26523ce66368aaf2caea16566/cuplaConfig.cmake#L102-L110

File: CMakeLists.txt
Comment (j-stephan) [reply CMakeLists.txt:70]: I looked into this a bit further and it needs to be fixed in alpaka itself. Right now alpaka's CMake installation script doesn't export the alpaka target. This allows us to reconfigure alpaka after installing it but comes with the issue we see here. I'll open an issue on the alpaka tracker to discuss this.

Comment (SimeonEhrig) [reply CMakeLists.txt]: I introduced a new target for internal linkage and removed the export of alpaka. Now, the linkage to alpaka is done, if the package was found in a external project. See commit  75eaa7edc6f8ff77169f1678d8b409a7db0b3cd1

**Comment by SimeonEhrig** [comment]
@j-stephan What do you think about the new target `interalvikunja` and the linkage on configure time?

**Comment by j-stephan** [comment]
Looks like a nice workaround for the problem. If this works with `add_subdirectory`, too, I'm okay with the merge.

**Pull request merged**
